[0.目录](../0.目录.md)
# 4.11 处理堆积错误
由于ZIO的错误类型是多态的，因此您可以使用错误通道来表示多个级别的故障。
这样可以简化一些其他复杂的代码。 一个示例是从数据库中查找用户概要文件。
现在，假设数据库查询可能失败，可能是因为用户未包含在数据库中。
但也可以说，在我们的数据库中，配置文件是可选的。
在这种情况下，我们的查找功能可能具有以下类型签名：

```
  trait DatabaseError
  trait UserProfile

  def lookupProfile(userId: String): ZIO[Any, DatabaseError, Option[UserProfile]] = ???
```

这种类型签名肯定没有错，如果我们只有一个持久性方法返回包装在Option中的结果，那么它可能就可以正常工作。
但是，如果我们发现自己处理的是包装在Option中的许多成功值，则ZIO的错误通道提供了一种更好的方法：我们可以“拆包”
Option等数据类型（以及其他面向失败的数据类型，包括Try，Either
，等等）将某些失败案例转移到错误渠道。
在下面的代码片段中，我们引入了一个新功能来执行此操作：

```
  def lookupProfile2(userId: String): ZIO[Any, Option[DatabaseError], UserProfile] = lookupProfile.foldM(
    error => ZIO.fail(Some(error)),
    success =>
      success match {
        case None => ZIO.fail(None)
        case Some(profile) => ZIO.succeed(profile)
      }
  )

```

此函数返回的效果可能因`Option
[DatabaseError]`而失败。这可能看起来像是一个奇怪的错误类型，但是如果您仔细考虑一下，这应该是有道理的：如果原始效果以“无”成功，则新效果将以“无”失败。但是，如果原始效果因错误e而失败，那么新效果将因Some（e）而失败。
将Option的失败案例转移到错误通道的新效果具有与原始效果相同的信息量。但是，新效果更易于使用，因为如果我们在其上调用flatMap（或将效果用于），则不必担心用户个人资料不在那里。相反，如果没有用户个人资料，则不会执行使用该用户个人资料的代码，我们可以在更高级别处理错误。
该技术适用于其他数据类型，例如Try和Either，本章中的一些练习要求您为这些情况编写帮助程序。
对于Option而言，这些帮助程序已经包含在ZIO中。您可以简单地在ZIO效果上调用某些选项，以将Option的None情况移至错误通道，并可选地将其移回成功通道。
例如，我们可以使用以下方法来实现lookupProfile2方法：

```
  def lookupProfile3(userId: String): ZIO[Any, Option[DatabaseError], UserProfile] = lookupProfile(userId).some
```

通过利用ZIO的类型化错误通道，您可以轻松处理包装的错误类型。
